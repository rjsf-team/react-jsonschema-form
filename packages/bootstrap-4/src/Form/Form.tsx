import { withTheme, FormProps } from "@rjsf/core";

import Theme from "../Theme";
import { StatelessComponent } from "react";

const Form:
  | React.ComponentClass<FormProps<any>>
  | StatelessComponent<FormProps<any>> = withTheme(Theme);

export default Form;
