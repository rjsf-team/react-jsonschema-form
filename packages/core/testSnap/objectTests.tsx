import { ComponentType } from 'react';
import renderer from 'react-test-renderer';
import { RJSFSchema, UiSchema } from '@rjsf/utils';
import validator from '@rjsf/validator-ajv8';

import { FormProps } from '../src';

const titleAndDesc = {
  title: 'Test field',
  description: 'a test description',
};

const uiTitleAndDesc: UiSchema = {
  'ui:options': {
    title: 'My Field',
    description: 'a fancier description',
  },
  a: {
    'ui:options': {
      title: 'My Item A',
      description: 'a fancier item A description',
    },
  },
  b: {
    'ui:options': {
      title: 'My Item B',
      description: 'a fancier item B description',
    },
  },
};

const labelsOff: UiSchema = {
  'ui:globalOptions': { label: false },
};

export default function arrayTests(Form: ComponentType<FormProps>) {
  describe('object fields', () => {
    test('object', () => {
      const schema: RJSFSchema = {
        type: 'object',
        properties: {
          a: { type: 'string', title: 'A' },
          b: { type: 'number', title: 'B' },
        },
      };
      const tree = renderer.create(<Form schema={schema} validator={validator} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
    test('additionalProperties', () => {
      const schema: RJSFSchema = {
        type: 'object',
        additionalProperties: true,
      };
      const tree = renderer.create(<Form schema={schema} validator={validator} formData={{ foo: 'foo' }} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
    test('show add button and fields if additionalProperties is true and not an object', () => {
      const schema: RJSFSchema = {
        additionalProperties: true,
      };
      const formData: any = {
        additionalProperty: 'should appear',
      };
      const tree = renderer.create(<Form schema={schema} validator={validator} formData={formData} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
    describe('with title and description', () => {
      test('object', () => {
        const schema: RJSFSchema = {
          type: 'object',
          ...titleAndDesc,
          properties: {
            a: { type: 'string', title: 'A', description: 'A description' },
            b: { type: 'number', title: 'B', description: 'B description' },
          },
        };
        const tree = renderer.create(<Form schema={schema} validator={validator} />).toJSON();
        expect(tree).toMatchSnapshot();
      });
      test('additionalProperties', () => {
        const schema: RJSFSchema = {
          type: 'object',
          ...titleAndDesc,
          additionalProperties: true,
        };
        const tree = renderer.create(<Form schema={schema} validator={validator} formData={{ foo: 'foo' }} />).toJSON();
        expect(tree).toMatchSnapshot();
      });
      test('show add button and fields if additionalProperties is true and not an object', () => {
        const schema: RJSFSchema = {
          ...titleAndDesc,
          additionalProperties: true,
        };
        const formData: any = {
          additionalProperty: 'should appear',
        };
        const tree = renderer.create(<Form schema={schema} validator={validator} formData={formData} />).toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
    describe('with title and description from uiSchema', () => {
      test('object', () => {
        const schema: RJSFSchema = {
          type: 'object',
          properties: {
            a: { type: 'string', title: 'A' },
            b: { type: 'number', title: 'B' },
          },
        };
        const tree = renderer.create(<Form schema={schema} uiSchema={uiTitleAndDesc} validator={validator} />).toJSON();
        expect(tree).toMatchSnapshot();
      });
      test('additionalProperties', () => {
        const schema: RJSFSchema = {
          type: 'object',
          additionalProperties: true,
        };
        const tree = renderer
          .create(<Form schema={schema} uiSchema={uiTitleAndDesc} validator={validator} formData={{ foo: 'foo' }} />)
          .toJSON();
        expect(tree).toMatchSnapshot();
      });
      test('show add button and fields if additionalProperties is true and not an object', () => {
        const schema: RJSFSchema = {
          additionalProperties: true,
        };
        const formData: any = {
          additionalProperty: 'should appear',
        };
        const tree = renderer
          .create(<Form schema={schema} uiSchema={uiTitleAndDesc} validator={validator} formData={formData} />)
          .toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
    describe('with title and description from both', () => {
      test('object', () => {
        const schema: RJSFSchema = {
          type: 'object',
          ...titleAndDesc,
          properties: {
            a: { type: 'string', title: 'A', description: 'A description' },
            b: { type: 'number', title: 'B', description: 'B description' },
          },
        };
        const tree = renderer.create(<Form schema={schema} uiSchema={uiTitleAndDesc} validator={validator} />).toJSON();
        expect(tree).toMatchSnapshot();
      });
      test('additionalProperties', () => {
        const schema: RJSFSchema = {
          type: 'object',
          ...titleAndDesc,
          additionalProperties: true,
        };
        const tree = renderer
          .create(<Form schema={schema} uiSchema={uiTitleAndDesc} validator={validator} formData={{ foo: 'foo' }} />)
          .toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
    describe('with title and description with global label off', () => {
      test('object', () => {
        const schema: RJSFSchema = {
          type: 'object',
          ...titleAndDesc,
          properties: {
            a: { type: 'string', title: 'A', description: 'A description' },
            b: { type: 'number', title: 'B', description: 'B description' },
          },
        };
        const tree = renderer.create(<Form schema={schema} uiSchema={labelsOff} validator={validator} />).toJSON();
        expect(tree).toMatchSnapshot();
      });
      test('additionalProperties', () => {
        const schema: RJSFSchema = {
          type: 'object',
          ...titleAndDesc,
          additionalProperties: true,
        };
        const tree = renderer
          .create(<Form schema={schema} uiSchema={labelsOff} validator={validator} formData={{ foo: 'foo' }} />)
          .toJSON();
        expect(tree).toMatchSnapshot();
      });
      test('show add button and fields if additionalProperties is true and not an object', () => {
        const schema: RJSFSchema = {
          ...titleAndDesc,
          additionalProperties: true,
        };
        const formData: any = {
          additionalProperty: 'should appear',
        };
        const tree = renderer
          .create(<Form schema={schema} uiSchema={labelsOff} validator={validator} formData={formData} />)
          .toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });
}
