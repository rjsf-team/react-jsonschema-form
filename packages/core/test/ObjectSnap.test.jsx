import Form from '../src';
import objectTests from '../testSnap/objectTests';

objectTests(Form);
