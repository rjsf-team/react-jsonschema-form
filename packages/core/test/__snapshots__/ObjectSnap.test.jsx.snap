// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`object fields additionalProperties 1`] = `
<form
  className="rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-object"
  >
    <fieldset
      id="root"
    >
      <div
        className="form-group field field-string"
      >
        <div
          className="row"
        >
          <div
            className="col-xs-5 form-additional"
          >
            <div
              className="form-group"
            >
              <label
                className="control-label"
                htmlFor="root_foo-key"
              >
                foo Key
              </label>
              <input
                className="form-control"
                defaultValue="foo"
                id="root_foo-key"
                onBlur={[Function]}
                type="text"
              />
            </div>
          </div>
          <div
            className="form-additional form-group col-xs-5"
          >
            <label
              className="control-label"
              htmlFor="root_foo"
            >
              foo
            </label>
            <input
              aria-describedby="root_foo__error root_foo__description root_foo__help"
              autoFocus={false}
              className="form-control"
              disabled={false}
              id="root_foo"
              label="foo"
              name="root_foo"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              readOnly={false}
              required={false}
              type="text"
              value="foo"
            />
          </div>
          <div
            className="col-xs-2"
          >
            <button
              className="btn btn-danger array-item-remove btn-block"
              disabled={false}
              onClick={[Function]}
              style={
                {
                  "border": "0",
                }
              }
              title="Remove"
              type="button"
            >
              <i
                className="glyphicon glyphicon-remove"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <p
          className="col-xs-3 col-xs-offset-9 text-right object-property-expand"
        >
          <button
            className="btn btn-info btn-add col-xs-12"
            disabled={false}
            onClick={[Function]}
            title="Add"
            type="button"
          >
            <i
              className="glyphicon glyphicon-plus"
            />
          </button>
        </p>
      </div>
    </fieldset>
  </div>
  <div>
    <button
      className="btn btn-info "
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`object fields object 1`] = `
<form
  className="rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-object"
  >
    <fieldset
      id="root"
    >
      <div
        className="form-group field field-string"
      >
        <label
          className="control-label"
          htmlFor="root_a"
        >
          A
        </label>
        <input
          aria-describedby="root_a__error root_a__description root_a__help"
          autoFocus={false}
          className="form-control"
          disabled={false}
          id="root_a"
          label="A"
          name="root_a"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder=""
          readOnly={false}
          required={false}
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group field field-number"
      >
        <label
          className="control-label"
          htmlFor="root_b"
        >
          B
        </label>
        <input
          aria-describedby="root_b__error root_b__description root_b__help"
          autoFocus={false}
          className="form-control"
          disabled={false}
          id="root_b"
          label="B"
          name="root_b"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder=""
          readOnly={false}
          required={false}
          step="any"
          type="number"
          value=""
        />
      </div>
    </fieldset>
  </div>
  <div>
    <button
      className="btn btn-info "
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`object fields show add button and fields if additionalProperties is true and not an object 1`] = `
<form
  className="rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-object"
  >
    <fieldset
      id="root"
    >
      <div
        className="form-group field field-string"
      >
        <div
          className="row"
        >
          <div
            className="col-xs-5 form-additional"
          >
            <div
              className="form-group"
            >
              <label
                className="control-label"
                htmlFor="root_additionalProperty-key"
              >
                additionalProperty Key
              </label>
              <input
                className="form-control"
                defaultValue="additionalProperty"
                id="root_additionalProperty-key"
                onBlur={[Function]}
                type="text"
              />
            </div>
          </div>
          <div
            className="form-additional form-group col-xs-5"
          >
            <label
              className="control-label"
              htmlFor="root_additionalProperty"
            >
              additionalProperty
            </label>
            <input
              aria-describedby="root_additionalProperty__error root_additionalProperty__description root_additionalProperty__help"
              autoFocus={false}
              className="form-control"
              disabled={false}
              id="root_additionalProperty"
              label="additionalProperty"
              name="root_additionalProperty"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              readOnly={false}
              required={false}
              type="text"
              value="should appear"
            />
          </div>
          <div
            className="col-xs-2"
          >
            <button
              className="btn btn-danger array-item-remove btn-block"
              disabled={false}
              onClick={[Function]}
              style={
                {
                  "border": "0",
                }
              }
              title="Remove"
              type="button"
            >
              <i
                className="glyphicon glyphicon-remove"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <p
          className="col-xs-3 col-xs-offset-9 text-right object-property-expand"
        >
          <button
            className="btn btn-info btn-add col-xs-12"
            disabled={false}
            onClick={[Function]}
            title="Add"
            type="button"
          >
            <i
              className="glyphicon glyphicon-plus"
            />
          </button>
        </p>
      </div>
    </fieldset>
  </div>
  <div>
    <button
      className="btn btn-info "
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`object fields with title and description additionalProperties 1`] = `
<form
  className="rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-object"
  >
    <fieldset
      id="root"
    >
      <legend
        id="root__title"
      >
        Test field
      </legend>
      <p
        className="field-description"
        id="root__description"
      >
        a test description
      </p>
      <div
        className="form-group field field-string"
      >
        <div
          className="row"
        >
          <div
            className="col-xs-5 form-additional"
          >
            <div
              className="form-group"
            >
              <label
                className="control-label"
                htmlFor="root_foo-key"
              >
                foo Key
              </label>
              <input
                className="form-control"
                defaultValue="foo"
                id="root_foo-key"
                onBlur={[Function]}
                type="text"
              />
            </div>
          </div>
          <div
            className="form-additional form-group col-xs-5"
          >
            <label
              className="control-label"
              htmlFor="root_foo"
            >
              foo
            </label>
            <input
              aria-describedby="root_foo__error root_foo__description root_foo__help"
              autoFocus={false}
              className="form-control"
              disabled={false}
              id="root_foo"
              label="foo"
              name="root_foo"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              readOnly={false}
              required={false}
              type="text"
              value="foo"
            />
          </div>
          <div
            className="col-xs-2"
          >
            <button
              className="btn btn-danger array-item-remove btn-block"
              disabled={false}
              onClick={[Function]}
              style={
                {
                  "border": "0",
                }
              }
              title="Remove"
              type="button"
            >
              <i
                className="glyphicon glyphicon-remove"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <p
          className="col-xs-3 col-xs-offset-9 text-right object-property-expand"
        >
          <button
            className="btn btn-info btn-add col-xs-12"
            disabled={false}
            onClick={[Function]}
            title="Add"
            type="button"
          >
            <i
              className="glyphicon glyphicon-plus"
            />
          </button>
        </p>
      </div>
    </fieldset>
  </div>
  <div>
    <button
      className="btn btn-info "
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`object fields with title and description from both additionalProperties 1`] = `
<form
  className="rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-object"
  >
    <fieldset
      id="root"
    >
      <legend
        id="root__title"
      >
        My Field
      </legend>
      <p
        className="field-description"
        id="root__description"
      >
        a fancier description
      </p>
      <div
        className="form-group field field-string"
      >
        <div
          className="row"
        >
          <div
            className="col-xs-5 form-additional"
          >
            <div
              className="form-group"
            >
              <label
                className="control-label"
                htmlFor="root_foo-key"
              >
                foo Key
              </label>
              <input
                className="form-control"
                defaultValue="foo"
                id="root_foo-key"
                onBlur={[Function]}
                type="text"
              />
            </div>
          </div>
          <div
            className="form-additional form-group col-xs-5"
          >
            <label
              className="control-label"
              htmlFor="root_foo"
            >
              foo
            </label>
            <input
              aria-describedby="root_foo__error root_foo__description root_foo__help"
              autoFocus={false}
              className="form-control"
              disabled={false}
              id="root_foo"
              label="foo"
              name="root_foo"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              readOnly={false}
              required={false}
              type="text"
              value="foo"
            />
          </div>
          <div
            className="col-xs-2"
          >
            <button
              className="btn btn-danger array-item-remove btn-block"
              disabled={false}
              onClick={[Function]}
              style={
                {
                  "border": "0",
                }
              }
              title="Remove"
              type="button"
            >
              <i
                className="glyphicon glyphicon-remove"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <p
          className="col-xs-3 col-xs-offset-9 text-right object-property-expand"
        >
          <button
            className="btn btn-info btn-add col-xs-12"
            disabled={false}
            onClick={[Function]}
            title="Add"
            type="button"
          >
            <i
              className="glyphicon glyphicon-plus"
            />
          </button>
        </p>
      </div>
    </fieldset>
  </div>
  <div>
    <button
      className="btn btn-info "
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`object fields with title and description from both object 1`] = `
<form
  className="rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-object"
  >
    <fieldset
      id="root"
    >
      <legend
        id="root__title"
      >
        My Field
      </legend>
      <p
        className="field-description"
        id="root__description"
      >
        a fancier description
      </p>
      <div
        className="form-group field field-string"
      >
        <label
          className="control-label"
          htmlFor="root_a"
        >
          My Item A
        </label>
        <p
          className="field-description"
          id="root_a__description"
        >
          a fancier item A description
        </p>
        <input
          aria-describedby="root_a__error root_a__description root_a__help"
          autoFocus={false}
          className="form-control"
          disabled={false}
          id="root_a"
          label="My Item A"
          name="root_a"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder=""
          readOnly={false}
          required={false}
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group field field-number"
      >
        <label
          className="control-label"
          htmlFor="root_b"
        >
          My Item B
        </label>
        <p
          className="field-description"
          id="root_b__description"
        >
          a fancier item B description
        </p>
        <input
          aria-describedby="root_b__error root_b__description root_b__help"
          autoFocus={false}
          className="form-control"
          disabled={false}
          id="root_b"
          label="My Item B"
          name="root_b"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder=""
          readOnly={false}
          required={false}
          step="any"
          type="number"
          value=""
        />
      </div>
    </fieldset>
  </div>
  <div>
    <button
      className="btn btn-info "
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`object fields with title and description from uiSchema additionalProperties 1`] = `
<form
  className="rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-object"
  >
    <fieldset
      id="root"
    >
      <legend
        id="root__title"
      >
        My Field
      </legend>
      <p
        className="field-description"
        id="root__description"
      >
        a fancier description
      </p>
      <div
        className="form-group field field-string"
      >
        <div
          className="row"
        >
          <div
            className="col-xs-5 form-additional"
          >
            <div
              className="form-group"
            >
              <label
                className="control-label"
                htmlFor="root_foo-key"
              >
                foo Key
              </label>
              <input
                className="form-control"
                defaultValue="foo"
                id="root_foo-key"
                onBlur={[Function]}
                type="text"
              />
            </div>
          </div>
          <div
            className="form-additional form-group col-xs-5"
          >
            <label
              className="control-label"
              htmlFor="root_foo"
            >
              foo
            </label>
            <input
              aria-describedby="root_foo__error root_foo__description root_foo__help"
              autoFocus={false}
              className="form-control"
              disabled={false}
              id="root_foo"
              label="foo"
              name="root_foo"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              readOnly={false}
              required={false}
              type="text"
              value="foo"
            />
          </div>
          <div
            className="col-xs-2"
          >
            <button
              className="btn btn-danger array-item-remove btn-block"
              disabled={false}
              onClick={[Function]}
              style={
                {
                  "border": "0",
                }
              }
              title="Remove"
              type="button"
            >
              <i
                className="glyphicon glyphicon-remove"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <p
          className="col-xs-3 col-xs-offset-9 text-right object-property-expand"
        >
          <button
            className="btn btn-info btn-add col-xs-12"
            disabled={false}
            onClick={[Function]}
            title="Add"
            type="button"
          >
            <i
              className="glyphicon glyphicon-plus"
            />
          </button>
        </p>
      </div>
    </fieldset>
  </div>
  <div>
    <button
      className="btn btn-info "
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`object fields with title and description from uiSchema object 1`] = `
<form
  className="rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-object"
  >
    <fieldset
      id="root"
    >
      <legend
        id="root__title"
      >
        My Field
      </legend>
      <p
        className="field-description"
        id="root__description"
      >
        a fancier description
      </p>
      <div
        className="form-group field field-string"
      >
        <label
          className="control-label"
          htmlFor="root_a"
        >
          My Item A
        </label>
        <p
          className="field-description"
          id="root_a__description"
        >
          a fancier item A description
        </p>
        <input
          aria-describedby="root_a__error root_a__description root_a__help"
          autoFocus={false}
          className="form-control"
          disabled={false}
          id="root_a"
          label="My Item A"
          name="root_a"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder=""
          readOnly={false}
          required={false}
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group field field-number"
      >
        <label
          className="control-label"
          htmlFor="root_b"
        >
          My Item B
        </label>
        <p
          className="field-description"
          id="root_b__description"
        >
          a fancier item B description
        </p>
        <input
          aria-describedby="root_b__error root_b__description root_b__help"
          autoFocus={false}
          className="form-control"
          disabled={false}
          id="root_b"
          label="My Item B"
          name="root_b"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder=""
          readOnly={false}
          required={false}
          step="any"
          type="number"
          value=""
        />
      </div>
    </fieldset>
  </div>
  <div>
    <button
      className="btn btn-info "
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`object fields with title and description from uiSchema show add button and fields if additionalProperties is true and not an object 1`] = `
<form
  className="rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-object"
  >
    <fieldset
      id="root"
    >
      <legend
        id="root__title"
      >
        My Field
      </legend>
      <p
        className="field-description"
        id="root__description"
      >
        a fancier description
      </p>
      <div
        className="form-group field field-string"
      >
        <div
          className="row"
        >
          <div
            className="col-xs-5 form-additional"
          >
            <div
              className="form-group"
            >
              <label
                className="control-label"
                htmlFor="root_additionalProperty-key"
              >
                additionalProperty Key
              </label>
              <input
                className="form-control"
                defaultValue="additionalProperty"
                id="root_additionalProperty-key"
                onBlur={[Function]}
                type="text"
              />
            </div>
          </div>
          <div
            className="form-additional form-group col-xs-5"
          >
            <label
              className="control-label"
              htmlFor="root_additionalProperty"
            >
              additionalProperty
            </label>
            <input
              aria-describedby="root_additionalProperty__error root_additionalProperty__description root_additionalProperty__help"
              autoFocus={false}
              className="form-control"
              disabled={false}
              id="root_additionalProperty"
              label="additionalProperty"
              name="root_additionalProperty"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              readOnly={false}
              required={false}
              type="text"
              value="should appear"
            />
          </div>
          <div
            className="col-xs-2"
          >
            <button
              className="btn btn-danger array-item-remove btn-block"
              disabled={false}
              onClick={[Function]}
              style={
                {
                  "border": "0",
                }
              }
              title="Remove"
              type="button"
            >
              <i
                className="glyphicon glyphicon-remove"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <p
          className="col-xs-3 col-xs-offset-9 text-right object-property-expand"
        >
          <button
            className="btn btn-info btn-add col-xs-12"
            disabled={false}
            onClick={[Function]}
            title="Add"
            type="button"
          >
            <i
              className="glyphicon glyphicon-plus"
            />
          </button>
        </p>
      </div>
    </fieldset>
  </div>
  <div>
    <button
      className="btn btn-info "
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`object fields with title and description object 1`] = `
<form
  className="rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-object"
  >
    <fieldset
      id="root"
    >
      <legend
        id="root__title"
      >
        Test field
      </legend>
      <p
        className="field-description"
        id="root__description"
      >
        a test description
      </p>
      <div
        className="form-group field field-string"
      >
        <label
          className="control-label"
          htmlFor="root_a"
        >
          A
        </label>
        <p
          className="field-description"
          id="root_a__description"
        >
          A description
        </p>
        <input
          aria-describedby="root_a__error root_a__description root_a__help"
          autoFocus={false}
          className="form-control"
          disabled={false}
          id="root_a"
          label="A"
          name="root_a"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder=""
          readOnly={false}
          required={false}
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group field field-number"
      >
        <label
          className="control-label"
          htmlFor="root_b"
        >
          B
        </label>
        <p
          className="field-description"
          id="root_b__description"
        >
          B description
        </p>
        <input
          aria-describedby="root_b__error root_b__description root_b__help"
          autoFocus={false}
          className="form-control"
          disabled={false}
          id="root_b"
          label="B"
          name="root_b"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder=""
          readOnly={false}
          required={false}
          step="any"
          type="number"
          value=""
        />
      </div>
    </fieldset>
  </div>
  <div>
    <button
      className="btn btn-info "
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`object fields with title and description show add button and fields if additionalProperties is true and not an object 1`] = `
<form
  className="rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-object"
  >
    <fieldset
      id="root"
    >
      <legend
        id="root__title"
      >
        Test field
      </legend>
      <p
        className="field-description"
        id="root__description"
      >
        a test description
      </p>
      <div
        className="form-group field field-string"
      >
        <div
          className="row"
        >
          <div
            className="col-xs-5 form-additional"
          >
            <div
              className="form-group"
            >
              <label
                className="control-label"
                htmlFor="root_additionalProperty-key"
              >
                additionalProperty Key
              </label>
              <input
                className="form-control"
                defaultValue="additionalProperty"
                id="root_additionalProperty-key"
                onBlur={[Function]}
                type="text"
              />
            </div>
          </div>
          <div
            className="form-additional form-group col-xs-5"
          >
            <label
              className="control-label"
              htmlFor="root_additionalProperty"
            >
              additionalProperty
            </label>
            <input
              aria-describedby="root_additionalProperty__error root_additionalProperty__description root_additionalProperty__help"
              autoFocus={false}
              className="form-control"
              disabled={false}
              id="root_additionalProperty"
              label="additionalProperty"
              name="root_additionalProperty"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              readOnly={false}
              required={false}
              type="text"
              value="should appear"
            />
          </div>
          <div
            className="col-xs-2"
          >
            <button
              className="btn btn-danger array-item-remove btn-block"
              disabled={false}
              onClick={[Function]}
              style={
                {
                  "border": "0",
                }
              }
              title="Remove"
              type="button"
            >
              <i
                className="glyphicon glyphicon-remove"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <p
          className="col-xs-3 col-xs-offset-9 text-right object-property-expand"
        >
          <button
            className="btn btn-info btn-add col-xs-12"
            disabled={false}
            onClick={[Function]}
            title="Add"
            type="button"
          >
            <i
              className="glyphicon glyphicon-plus"
            />
          </button>
        </p>
      </div>
    </fieldset>
  </div>
  <div>
    <button
      className="btn btn-info "
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`object fields with title and description with global label off additionalProperties 1`] = `
<form
  className="rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-object"
  >
    <fieldset
      id="root"
    >
      <div
        className="form-group field field-string"
      >
        <div
          className="row"
        >
          <div
            className="col-xs-5 form-additional"
          >
            <div
              className="form-group"
            >
              <label
                className="control-label"
                htmlFor="root_foo-key"
              >
                foo Key
              </label>
              <input
                className="form-control"
                defaultValue="foo"
                id="root_foo-key"
                onBlur={[Function]}
                type="text"
              />
            </div>
          </div>
          <div
            className="form-additional form-group col-xs-5"
          >
            <input
              aria-describedby="root_foo__error root_foo__description root_foo__help"
              autoFocus={false}
              className="form-control"
              disabled={false}
              id="root_foo"
              label="foo"
              name="root_foo"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              readOnly={false}
              required={false}
              type="text"
              value="foo"
            />
          </div>
          <div
            className="col-xs-2"
          >
            <button
              className="btn btn-danger array-item-remove btn-block"
              disabled={false}
              onClick={[Function]}
              style={
                {
                  "border": "0",
                }
              }
              title="Remove"
              type="button"
            >
              <i
                className="glyphicon glyphicon-remove"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <p
          className="col-xs-3 col-xs-offset-9 text-right object-property-expand"
        >
          <button
            className="btn btn-info btn-add col-xs-12"
            disabled={false}
            onClick={[Function]}
            title="Add"
            type="button"
          >
            <i
              className="glyphicon glyphicon-plus"
            />
          </button>
        </p>
      </div>
    </fieldset>
  </div>
  <div>
    <button
      className="btn btn-info "
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`object fields with title and description with global label off object 1`] = `
<form
  className="rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-object"
  >
    <fieldset
      id="root"
    >
      <div
        className="form-group field field-string"
      >
        <input
          aria-describedby="root_a__error root_a__description root_a__help"
          autoFocus={false}
          className="form-control"
          disabled={false}
          id="root_a"
          label="A"
          name="root_a"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder=""
          readOnly={false}
          required={false}
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group field field-number"
      >
        <input
          aria-describedby="root_b__error root_b__description root_b__help"
          autoFocus={false}
          className="form-control"
          disabled={false}
          id="root_b"
          label="B"
          name="root_b"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder=""
          readOnly={false}
          required={false}
          step="any"
          type="number"
          value=""
        />
      </div>
    </fieldset>
  </div>
  <div>
    <button
      className="btn btn-info "
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`object fields with title and description with global label off show add button and fields if additionalProperties is true and not an object 1`] = `
<form
  className="rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-object"
  >
    <fieldset
      id="root"
    >
      <div
        className="form-group field field-string"
      >
        <div
          className="row"
        >
          <div
            className="col-xs-5 form-additional"
          >
            <div
              className="form-group"
            >
              <label
                className="control-label"
                htmlFor="root_additionalProperty-key"
              >
                additionalProperty Key
              </label>
              <input
                className="form-control"
                defaultValue="additionalProperty"
                id="root_additionalProperty-key"
                onBlur={[Function]}
                type="text"
              />
            </div>
          </div>
          <div
            className="form-additional form-group col-xs-5"
          >
            <input
              aria-describedby="root_additionalProperty__error root_additionalProperty__description root_additionalProperty__help"
              autoFocus={false}
              className="form-control"
              disabled={false}
              id="root_additionalProperty"
              label="additionalProperty"
              name="root_additionalProperty"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              readOnly={false}
              required={false}
              type="text"
              value="should appear"
            />
          </div>
          <div
            className="col-xs-2"
          >
            <button
              className="btn btn-danger array-item-remove btn-block"
              disabled={false}
              onClick={[Function]}
              style={
                {
                  "border": "0",
                }
              }
              title="Remove"
              type="button"
            >
              <i
                className="glyphicon glyphicon-remove"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <p
          className="col-xs-3 col-xs-offset-9 text-right object-property-expand"
        >
          <button
            className="btn btn-info btn-add col-xs-12"
            disabled={false}
            onClick={[Function]}
            title="Add"
            type="button"
          >
            <i
              className="glyphicon glyphicon-plus"
            />
          </button>
        </p>
      </div>
    </fieldset>
  </div>
  <div>
    <button
      className="btn btn-info "
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
